core/src/main/java/org/apache/accumulo/core/spi/common/ContextClassLoaderFactory.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.core.spi.common;
+
+import java.util.Map;
+import java.util.function.Supplier;
+
+/**
+ * The ClassLoaderFactory is defined by the property general.context.factory. The factory
+ * implementation is configured externally to Accumulo and will return a ClassLoader for a given
+ * contextName.
+ *
+ * @since 2.1.0
+ *
+ */
+public interface ContextClassLoaderFactory {
+
+  /**
+   * Initialize the ClassLoaderFactory. Implementations may need a reference to the configuration so
+   * that it can clean up contexts that are no longer being used.
+   *
+   * @param contextProperties
+   *          Accumulo configuration properties
+   * @throws Exception
+   *           if error initializing ClassLoaderFactory
+   */
+  void initialize(Supplier<Map<String,String>> contextProperties) throws Exception;
```

Comment:
The following are the rational behind the suggestion below.
 * Using a parameters interface makes evolving the SPI over time much simpler.  Its easy to add additional parameters in the future w/o breaking existing code that was written against the SPI.
 * Using the [Configuration](https://github.com/apache/accumulo/blob/353c611891a8b88b307dcba5e884467d4191b1f2/core/src/main/java/org/apache/accumulo/core/spi/common/ServiceEnvironment.java#L40) interface used by other SPI interface is consistent and offers more options.  For example this interface offers an isSet() method that allows checking if a user set a property.  There is no way to do this by analyzing the values because some props have default values.

```suggestion
  public interface InitParameters {
      /**
   * @return A view of Accumulo's system level configuration. This is backed by system level config
   *         in zookeeper, which falls back to site configuration, which falls back to the default
   *         configuration.
   */
    Configuration getConfiguration();
  }

  /**
   * Initialize the ClassLoaderFactory. Implementations may need a reference to the configuration so
   * that it can clean up contexts that are no longer being used.
   *
   * @throws Exception
   *           if error initializing ClassLoaderFactory
   */
  void initialize(InitParameters initParam) throws Exception;
```

Reply:
@keith-turner It looks like ServiceEnvironmentImpl and ConfigurationImpl are the classes used to create the Configuration used in this manner. Both of these objects are in server-base, which depends on accumulo-core. Suggestions on how to move forward? I think ContextClassLoaderFactory should remain in core as that is where the SPI classes are. It looks like I might be able to move ServiceEnvironmentImpl up to accumulo-core.

Reply:
I can't move ServiceEnvironmentImpl up to accumulo-core due to its depedency on ServerContext.

Reply:
I think what you did with movin ConfigurationImpl and using the ConfigurationImpl is good.  Would still recommend the parameters interface as it gives more options for mutating the SPI in the future w/o causing pain to anyone who has taken the time to implement the interface.

Reply:
> Would still recommend the parameters interface as it gives more options for mutating the SPI in the future w/o causing pain to anyone who has taken the time to implement the interface.

I think I would disagree with the need to create an evolve-able parameter type here. I don't think it's a good idea to support a dedicated parameter type because that implies we're constructing a framework that can evolve, on which complex subclass implementations can support being initialized based on internal Accumulo objects. That's antithetical to what we're trying to do here.

This ClassloaderFactory is not your typical SPI (like sampling or iterators or balancers), where behavior is coupled to Accumulo functionality. Rather, classloading is something that should be done largely independently of Accumulo... as it is something on which other parts of Accumulo depend. Its initialization API should not itself depend on Accumulo objects that would be passed in using these parameter types, because that could create cyclical dependencies (Accumulo depends on classloader, which depends on Accumulo internal objects passed as parameters, which depends on classloader...).

The only thing this interface needs from Accumulo is initial configuration, which just happens to be carried in Accumulo's own configuration files. We should merely pass that information along to this. We should never pass along any other internal state objects as parameters. Besides, the classloader factory should be instantiated *early*, before any meaningful state is generated in Accumulo. So, it should never have the opportunity to initialize based on internal Accumulo state anyway... just initial configuration files.

So, I think it's a conscious decision to make the initialization parameter only a configuration mapping. It will ensure we keep the classloading stuff as decoupled as possible from Accumulo's internal workings, and merely responsible for providing classloaders that the rest of Accumulo can reliably depend on, based on the implementer's own independent criteria.

Reply:
Actually, because classloading can get messy, I've debated arguing in favor of removing this initialize method entirely, and just expecting implementations rely on their own way of getting configuration (their own config files, or environment parsing), so it's even more decoupled from Accumulo. But... that would make our own provided implementations that are currently built-in (for preserving existing behavior) more difficult, because they wouldn't be able to read `general.classpaths`, etc.

Reply:
The concept that we know everything we may ever want to pass into a method is not something that has worked out too well in the past for the Accumulo API and its riddled with many duplicative overloaded methods as a result.   The parameter interface is an attempt to avoid repeating a mistake that has been made many times in the past.   

I don't think there is a 0% chance that another parameter will ever be needed.  While valid use case likely exists, we may or may not encounter them.  I have no idea what the chances are, so I am ok w/ someone making the judgement call that its not worth worrying about.  Personally I would worry about it simply based on past experience w/ API evolution.

Reply:
It wasn't so much that I think "we know everything we may ever want to pass", so much as it was "I want to restrict what we can pass, to avoid scope creep". I don't feel too strongly about this and am okay with Keith's API strategy here.